qiskit_nature/circuit/library/ansatzes/uvcc.py

```diff
@@ -0,0 +1,242 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+""" The Unitary Vibrational Coupled-Cluster Single and Double excitations variational form. """
+
+from functools import partial
+from typing import Any, Callable, List, Optional, Tuple, Union
+
+import logging
+
+from qiskit.circuit import QuantumCircuit
+from qiskit.opflow import PauliTrotterEvolution
+from qiskit_nature import QiskitNatureError
+from qiskit_nature.operators.second_quantization import SecondQuantizedOp, VibrationalOp
+from qiskit_nature.operators.second_quantization.qubit_converter import QubitConverter
+from .evolved_operator_ansatz import EvolvedOperatorAnsatz
+from .utils.vibration_excitation_generator import generate_vibration_excitations
+
+logger = logging.getLogger(__name__)
+
+
+class UVCC(EvolvedOperatorAnsatz):
+    """
+    This trial wavefunction is a Unitary Vibrational Coupled-Cluster Single and Double excitations
+    variational form.
+    For more information, see Ollitrault Pauline J., Chemical science 11 (2020): 6842-6855.
+    """
+
+    EXCITATION_TYPE = {
+        's': 1,
+        'd': 2,
+        't': 3,
+        'q': 4,
+    }
+
+    def __init__(self,
+                 qubit_converter: Optional[QubitConverter] = None,
+                 num_modals: Optional[List[int]] = None,
+                 excitations: Optional[Union[str, int, List[int], Callable[
+                     [int, Tuple[int, int]], List[Tuple[Tuple[Any, ...], ...]]]]] = None,
+                 reps: int = 1,
+                 initial_state: Optional[QuantumCircuit] = None):
+        """
+
+        Args:
+            qubit_converter: the QubitConverter instance which takes care of mapping a
+            :class:`~.SecondQuantizedOp` to a :class:`~.PauliSumOp` as well as performing all
+            configured symmetry reductions on it.
+            num_modals: Is a list defining the number of modals per mode. E.g. for a 3 modes system
+                with 4 modals per mode num_modals = [4,4,4]
+            excitations: this can be any of the following:
+                - a `str` which contains the types of excitations. Allowed characters are:
+                    - `s` for singles
+                    - `d` for doubles
+                    - `t` for triples
+                    - `q` for quadruples
+                - a single, positive `int` which denotes the number of excitations (1 == `s`, etc.)
+                - a list of positive integers
+                - a `callable` object which is used to generate the excitations. The `callable` must
+                  take the following __keyword__ arguments:
+                      - `num_modals`: the same as above
+                  and must return a `List[Tuple[Tuple[Any, ...], ...]]`.
+            reps: number of replica of basic module
```

Comment:
I think this short for 'repetitions' rather than replica, so it may be preferable to say number of repetitions